Feature/mom6 tools

jkbk2004 opened the pull request:
## Description
initial inclusion of mom6-tools for post-processing plot option


### Issue(s) addressed
post-processing plot option is not included in workflow. But this pr will facilitate the post-processing option later. Standalone python scripts in src/mom6-tools.plot were tested to make plots with cice5 and mom6 diagnostic outputs. Short how-to-use instruction can be found README in the directory

jkbk2004 committed:
first commit to include mom6-tools

jkbk2004 committed:
edited readme

jkbk2004 committed:
editted more on readme

jkbk2004 committed:
more correction on readme

jkbk2004 committed:
more correction on readme

jkbk2004 committed:
updated mom6-tools path

flampouris reviewed (commented):
@jkbk2004 short question: Do we have a local repository for the mom6tools or do they just live in "common"?

jkbk2004 commented:
> @jkbk2004 short question: Do we have a local repository for the mom6tools or do they just live in "common"?
installation
@flampouris we have https://github.com/NOAA-EMC/mom6-tools. But I want to keep it on common until I figure out what to do with installation issue on hera and cartopy local shapefile issue.

jkbk2004 commented:
> > @jkbk2004 short question: Do we have a local repository for the mom6tools or do they just live in "common"?
> > installation
> > @flampouris we have https://github.com/NOAA-EMC/mom6-tools. But I want to keep it on common until I figure out what to do with installation issue on hera and cartopy local shapefile issue.

@flampouris maybe the tool will get adjusted and we will make more drivers until then.

ShastriPaturi commented:
@jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
An error occurs on step 0-1 as : Permission denied

jkbk2004 commented:
> @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> An error occurs on step 0-1 as : Permission denied

@ShastriPaturi Try now.

ShastriPaturi commented:
> > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > An error occurs on step 0-1 as : Permission denied
> 
> @ShastriPaturi Try now.

@jkbk2004, same error
cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied

jkbk2004 commented:
> > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > An error occurs on step 0-1 as : Permission denied
> > 
> > 
> > @ShastriPaturi Try now.
> 
> @jkbk2004, same error
> cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied

@ShastriPaturi ok, I did -R 777 now.

ShastriPaturi commented:
> > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > An error occurs on step 0-1 as : Permission denied
> > > 
> > > 
> > > @ShastriPaturi Try now.
> > 
> > 
> > @jkbk2004, same error
> > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> 
> @ShastriPaturi ok, I did -R 777 now.

I can't ls also:
ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
ls: cannot open directory /home/Jong.Kim/: Permission denied

jkbk2004 commented:
> > > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > > An error occurs on step 0-1 as : Permission denied
> > > > 
> > > > 
> > > > @ShastriPaturi Try now.
> > > 
> > > 
> > > @jkbk2004, same error
> > > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> > 
> > 
> > @ShastriPaturi ok, I did -R 777 now.
> 
> I can't ls also:
> ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
> ls: cannot open directory /home/Jong.Kim/: Permission denied

@ShastriPaturi Oh! my home directory. Can you try now?

ShastriPaturi commented:
> > > > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > > > An error occurs on step 0-1 as : Permission denied
> > > > > 
> > > > > 
> > > > > @ShastriPaturi Try now.
> > > > 
> > > > 
> > > > @jkbk2004, same error
> > > > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> > > 
> > > 
> > > @ShastriPaturi ok, I did -R 777 now.
> > 
> > 
> > I can't ls also:
> > ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
> > ls: cannot open directory /home/Jong.Kim/: Permission denied
> 
> @ShastriPaturi Oh! my home directory. Can you try now?

Same error: ls: cannot open directory /home/Jong.Kim/: Permission denied

ShastriPaturi commented:
> > > > > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > > > > An error occurs on step 0-1 as : Permission denied
> > > > > > 
> > > > > > 
> > > > > > @ShastriPaturi Try now.
> > > > > 
> > > > > 
> > > > > @jkbk2004, same error
> > > > > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> > > > 
> > > > 
> > > > @ShastriPaturi ok, I did -R 777 now.
> > > 
> > > 
> > > I can't ls also:
> > > ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
> > > ls: cannot open directory /home/Jong.Kim/: Permission denied
> > 
> > 
> > @ShastriPaturi Oh! my home directory. Can you try now?
> 
> Same error: ls: cannot open directory /home/Jong.Kim/: Permission denied

Can you put the shapefiles in marineda? It might be easier to copy.

jkbk2004 commented:
> > > > > > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > > > > > An error occurs on step 0-1 as : Permission denied
> > > > > > > 
> > > > > > > 
> > > > > > > @ShastriPaturi Try now.
> > > > > > 
> > > > > > 
> > > > > > @jkbk2004, same error
> > > > > > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> > > > > 
> > > > > 
> > > > > @ShastriPaturi ok, I did -R 777 now.
> > > > 
> > > > 
> > > > I can't ls also:
> > > > ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
> > > > ls: cannot open directory /home/Jong.Kim/: Permission denied
> > > 
> > > 
> > > @ShastriPaturi Oh! my home directory. Can you try now?
> > 
> > 
> > Same error: ls: cannot open directory /home/Jong.Kim/: Permission denied
> 
> Can you put the shapefiles in marineda? It might be easier to copy.

@ShastriPaturi I put /common/cartopy. I have to change the way I use cartopy. Maybe I have to use cartopy.config. Let me test and I put it in this pr.

ShastriPaturi commented:
> > > > > > > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > > > > > > An error occurs on step 0-1 as : Permission denied
> > > > > > > > 
> > > > > > > > 
> > > > > > > > @ShastriPaturi Try now.
> > > > > > > 
> > > > > > > 
> > > > > > > @jkbk2004, same error
> > > > > > > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> > > > > > 
> > > > > > 
> > > > > > @ShastriPaturi ok, I did -R 777 now.
> > > > > 
> > > > > 
> > > > > I can't ls also:
> > > > > ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
> > > > > ls: cannot open directory /home/Jong.Kim/: Permission denied
> > > > 
> > > > 
> > > > @ShastriPaturi Oh! my home directory. Can you try now?
> > > 
> > > 
> > > Same error: ls: cannot open directory /home/Jong.Kim/: Permission denied
> > 
> > 
> > Can you put the shapefiles in marineda? It might be easier to copy.
> 
> @ShastriPaturi I put /common/cartopy. I have to change the way I use cartopy. Maybe I have to use cartopy.config. Let me test and I put it in this pr.

@jkbk2004 I will wait for your command, Sir!!!

jkbk2004 committed:
moved cartopused cartopy.config to use common shapefiles

jkbk2004 commented:
> > > > > > > > > > @jkbk2004, please open your folder /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*' for copying the files
> > > > > > > > > > An error occurs on step 0-1 as : Permission denied
> > > > > > > > > 
> > > > > > > > > 
> > > > > > > > > @ShastriPaturi Try now.
> > > > > > > > 
> > > > > > > > 
> > > > > > > > @jkbk2004, same error
> > > > > > > > cp: cannot stat '/home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/*': Permission denied
> > > > > > > 
> > > > > > > 
> > > > > > > @ShastriPaturi ok, I did -R 777 now.
> > > > > > 
> > > > > > 
> > > > > > I can't ls also:
> > > > > > ls: cannot access /home/Jong.Kim/.local/share/cartopy/shapefiles/natural_earth/physical/: Permission denied
> > > > > > ls: cannot open directory /home/Jong.Kim/: Permission denied
> > > > > 
> > > > > 
> > > > > @ShastriPaturi Oh! my home directory. Can you try now?
> > > > 
> > > > 
> > > > Same error: ls: cannot open directory /home/Jong.Kim/: Permission denied
> > > 
> > > 
> > > Can you put the shapefiles in marineda? It might be easier to copy.
> > 
> > 
> > @ShastriPaturi I put /common/cartopy. I have to change the way I use cartopy. Maybe I have to use cartopy.config. Let me test and I put it in this pr.
> 
> @jkbk2004 I will wait for your command, Sir!!!

@ShastriPaturi I updated ice.plot.py to use cartopy.config for common shapefiles. updated readme as well. you don't need to copy shapefiles anymore. We just need to point to location of files inside script. we don't need cartopy for regular surface plots. only with polar project so far.

ShastriPaturi commented:
@jkbk2004 where is the location of ocean_geometry.nc?
For soca, the files is now soca_gridspec.nc. It is in the same directory as the output files.

However, while using grid_spec.nc, I get the following error:
Traceback (most recent call last):
  File "ice.plot.py", line 35, in <module>
    grd.area_t=grd.Ah
AttributeError: type object 'MOM6_grd' has no attribute 'Ah'

jkbk2004 commented:
@ShastriPaturi I have to update diag_table to dump right variables to use the tool. Meanwhile, you can test the tool with ocean_geometry.nc file I put in common directory. Once we get right sets of all dia variables, I will clean up common directory.

ShastriPaturi commented:
For sfc.time.plot.py, Are the plots specific for 2012 data?
In Lines 27-29, is it hard coded: 
# initial and final years for computing time mean
year_start = 3
year_end = 195
xlabel='hours since 2012-01-01 00'

Also the script fails with the following error:
Traceback (most recent call last):
  File "sfc.time.plot.py", line 51, in <module>
    time_mean_latlon(args,grd)
  File "/scratch2/NCEPDEV/marineda/common/mom6-tools/mom6_tools/latlon_analysis.py", line 164, in time_mean_latlon
    save=filename)
  File "/scratch2/NCEPDEV/marineda/common/mom6-tools/mom6_tools/m6plot.py", line 593, in xyplot
    if save is not None: plt.savefig(save)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/pyplot.py", line 722, in savefig
    res = fig.savefig(*args, **kwargs)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/figure.py", line 2180, in savefig
    self.canvas.print_figure(fname, **kwargs)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 88, in print_figure
    super().print_figure(*args, **kwargs)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/backend_bases.py", line 2082, in print_figure
    **kwargs)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 530, in print_png
    cbook.open_file_cm(filename_or_obj, "wb") as fh:
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/cbook/__init__.py", line 447, in open_file_cm
    fh, opened = to_filehandle(path_or_file, mode, True, encoding)
  File "/scratch2/NCEPDEV/marine/Todd.Spindler/save/anaconda3-standard/lib/python3.7/site-packages/matplotlib/cbook/__init__.py", line 432, in to_filehandle
    fh = open(fname, flag, encoding=encoding)
FileNotFoundError: [Errno 2] No such file or directory: 'PNG/SST.png'

This is the command being used:
python sfc.time.plot.py -grid /scratch2/NCEPDEV/marineda/common/ocean_geometry.nc -data /scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/workflowtest_QCexp_SST25/2011100112/fcst/ocn_2011_10_02_03.nc
same as for sfc.plot.py.

jkbk2004 commented:
@ShastriPaturi agree! its hard coded. Let me update the script. I think I have to put more specific instruction in readme as well.

jkbk2004 committed:
updated sfc.time average script: less hard coded

jkbk2004 committed:
updated readme

jkbk2004 committed:
updated readme

jkbk2004 commented:
@ShastriPaturi I updated mom6-tools and sfc.time.plot.py. Its less hard coded now. Go ahead to test.

ShastriPaturi commented:
> @ShastriPaturi I updated mom6-tools and sfc.time.plot.py. Its less hard coded now. Go ahead to test.

@jkbk2004, the command:
python sfc.time.plot.py -grid /scratch2/NCEPDEV/marineda/common/ocean_geometry.nc -data /scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/workflowtest_QCexp_SST25/2011100112/fcst/ocn_2011_10_02_03.nc
creates the png files for SST and SSH time average plots in a directory time_mean. however, the area-average statistic time-series plots are empty.
When I use the data as ocn_2011_10_02_03.nc, does it open the ocn_2011_10_01_15.nc and ocn_2011_10_01_21.nc to calculate the mean and the statistics?
The title of the surface average plot says hours since 2010-11-01 00:00. Our DA window is at 12 hrs. Shouldn't the averaging and the title be reflective of that?

Also, when I run the command the second time, I get 
==> {} has been saved, moving to the next one.

jkbk2004 commented:
@ShastriPaturi -data input for time average should be multiple files in time series. Something like ocn_2011_10_0[1-2]_*.nc

jkbk2004 commented:
@ShastriPaturi ref_date in the script is used to set time scale unit in x-axis. If you need different ref_date, you can change it inside script. script counts from ref_date for time series plot. Once the scripts settles, we can come up with a better way to set.

ShastriPaturi commented:
> @ShastriPaturi -data input for time average should be multiple files in time series. Something like ocn_2011_10_0[1-2]_*.nc

@jkbk2004, I have tried that and fails because the time_mean directory is created inside the RUNCDATE/fcst directory where the ocn_2011_10_0[1-2]_*.nc reside. The time_mean directory should be two level up from the fcst directory.

And now hera is down.

jkbk2004 committed:
added figs_path and var options

jkbk2004 commented:
@ShastriPaturi I added -figs_path input argument. You can specify the path you want to save with -figs_path: something like -figs_path /scratch2/NCEPDEV/marineda/Shastri.Paturi/Shastri.Paturi/rundir/time_mean

jkbk2004 committed:
updated date parse and added time stamp on time-avg fig filenames

jkbk2004 commented:
@ShastriPaturi I updated mom6-tools to include time stamp on time average figure files. I confirm that sfc.time.plot.py can handle multiple directories. So, go ahead to update the branch and test. Use -figs_path option to save on the location you want.

ShastriPaturi commented on src/mom6-tools.plot/README.md:
```diff
@@ -0,0 +1,31 @@
+
+-figs_path (optional): path to save png files (e.g., -figs_path ./fcst)
+
+-var (optional): variable names to plot (e.g., -var hice aice or -var hi_h aice_h)
```
Are there any variables as hi_h and aice_h?
These vaiables were not found in cic.socagodas.an.2011-10-01T12:00:00Z.nc while testing.

Is this a typo?

ShastriPaturi reviewed (approved):
The PR has been tested and produces the results as promised.
The README.md on line 13 has (e.g., -var hice aice or -var hi_h aice_h). Are there any ice variables as hi_h and acice_h or is it a typo? While testing The variables hi_h and acice_h were not found in the cic.socagodas.an.2011-10-01T12:00:00Z.nc file.
Looks good, otherwise.

jkbk2004 commented:
@ShastriPaturi hi_h aice_h seem to be used in history files. So, I just mentioned in readme. But by default, script works with aice/hice.

flampouris commented:
Merged. 

Thank you @jkbk2004  and @ShastriPaturi for your excellent job, one milestone less.
